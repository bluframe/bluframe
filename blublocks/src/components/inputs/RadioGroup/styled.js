// @flow
/* eslint-disable no-ternary */

import styled, { type StyledComponent } from "styled-components"
import type { Theme } from "@bluframe/blublocks"

export const Wrapper: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  display: grid;
  row-gap: 8px;
`

export const Label: StyledComponent<
  {||},
  Theme,
  HTMLLabelElement
> = styled.label`
  color: ${({ theme }) => theme.palette.text.main};
  font-size: 16px;
`

export const Options: StyledComponent<{||}, Theme, HTMLDivElement> = styled.div`
  display: grid;
  row-gap: 4px;
`
