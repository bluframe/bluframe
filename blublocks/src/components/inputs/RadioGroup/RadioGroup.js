// @flow

import { Label, Options, Wrapper } from "./styled"
import type { ComponentProps } from "."
import Radio from "components/inputs/Radio"
import React from "react"

const RadioGroup = ({
  className,
  groupId,
  label,
  labelId,
  name,
  onChange,
  options,
  selected
}: ComponentProps): React$Node => (
  <Wrapper
    aria-labelledby={labelId}
    className={className}
    id={groupId}
    role="radiogroup"
  >
    <Label id={labelId}>{label}</Label>
    <Options>
      {options.map((option) => (
        <Radio
          {...option}
          inputId={`${groupId}-${option.value}`}
          isChecked={option.value === selected}
          key={option.value}
          name={name}
          onChange={onChange}
        />
      ))}
    </Options>
  </Wrapper>
)

export default RadioGroup
