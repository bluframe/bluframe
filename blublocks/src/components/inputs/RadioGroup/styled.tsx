/* eslint-disable no-ternary */

import styled from "styled-components"
import { Theme } from "styles/theme"

export const Wrapper = styled.div`
  display: grid;
  row-gap: 8px;
`

export const Label = styled.label`
  color: ${({ theme }: { theme: Theme }) => theme.palette.text.main};
  font-size: 16px;
`

export const Options = styled.div`
  display: grid;
  row-gap: 4px;
`
