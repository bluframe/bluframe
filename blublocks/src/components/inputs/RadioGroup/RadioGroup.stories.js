// @flow

import RadioGroup, { type Props } from "components/inputs/RadioGroup"
import React from "react"
import type { Story } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const stories = {
  component: RadioGroup,
  title: "Inputs/RadioGroup"
}

const Template = (args: Props) => <RadioGroup {...args} />

const args = {
  label: "Select a fruit",
  name: "fruit",
  onChange: action("onChange"),
  options: [
    {
      label: "Apple",
      value: "apple"
    },
    {
      label: "Banana",
      value: "banana"
    }
  ],
  selected: null
}

export const Default: Story<Props> = Template.bind({})
Default.args = args

export const Selected: Story<Props> = Template.bind({})
Selected.args = {
  ...args,
  selected: "banana"
}

export default stories
