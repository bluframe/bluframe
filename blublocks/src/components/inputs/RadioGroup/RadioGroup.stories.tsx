import RadioGroup from "components/inputs/RadioGroup"
import { StoryObj } from "@storybook/react"
import { action } from "@storybook/addon-actions"

const meta = {
  component: RadioGroup,
  title: "Inputs/RadioGroup"
}

type Story = StoryObj<typeof RadioGroup>

const args = {
  label: "Select a fruit",
  name: "fruit",
  onChange: action("onChange"),
  options: [
    {
      label: "Apple",
      value: "apple"
    },
    {
      label: "Banana",
      value: "banana"
    }
  ],
  selected: null
}

export const Default: Story = {
  args
}

export const Selected: Story = {
  args: {
    ...args,
    selected: "banana"
  }
}

export default meta
