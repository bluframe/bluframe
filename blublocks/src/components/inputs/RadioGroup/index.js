// @flow

import RadioGroup from "./RadioGroup"
import type { RadioGroupProps } from "@bluframe/blublocks"
import { prepareComponent } from "@bluframe/grapple"
import { useMemo } from "react"

export type Props = {|
  ...RadioGroupProps
|}

export type ComponentProps = {|
  ...Props,
  +groupId: string,
  +labelId: string
|}

const FIRST = 0

const usePrepareComponent = (props: Props): ComponentProps => {
  const groupId = `${props.name}-radiogroup`
  const labelId = `${groupId}-label`

  // Auto select first option if no selected option is provided
  const selected = useMemo(() => {
    if (props.selected) {
      return props.selected
    }

    if (!props.options.length) {
      return null
    }

    return props.options[FIRST].value
  }, [props.options, props.selected])

  return {
    ...props,
    groupId,
    labelId,
    selected
  }
}

export default (prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  RadioGroup
): React$ComponentType<Props>)
