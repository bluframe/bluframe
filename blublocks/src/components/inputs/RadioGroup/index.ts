import React, { useMemo } from "react"
import RadioGroup from "./RadioGroup"
import { prepareComponent } from "@bluframe/grapple"

export type OptionValue = number | string

export interface Option {
  label: React.ReactNode
  value: OptionValue
}

export interface Props {
  className?: string
  label: React.ReactNode
  name: string
  // eslint-disable-next-line no-unused-vars
  onChange: (value: OptionValue) => void
  options: Option[]
  selected: OptionValue | null
}

export interface ComponentProps extends Props {
  groupId: string
  labelId: string
}

const FIRST = 0

const usePrepareComponent = (props: Props): ComponentProps => {
  const groupId = `${props.name}-radiogroup`
  const labelId = `${groupId}-label`

  // Auto select first option if no selected option is provided
  const selected = useMemo(() => {
    if (props.selected) {
      return props.selected
    }

    if (!props.options.length) {
      return null
    }

    return props.options[FIRST].value
  }, [props.options, props.selected])

  return {
    ...props,
    groupId,
    labelId,
    selected
  }
}

export default prepareComponent<Props, ComponentProps>(usePrepareComponent)(
  RadioGroup
)
