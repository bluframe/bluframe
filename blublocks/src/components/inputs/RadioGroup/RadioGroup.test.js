// @flow

import type { ComponentProps } from "."
import RadioGroup from "./RadioGroup"
import React from "react"
import { render } from "tests"

jest.mock(".", () => ({}))
jest.mock("components/inputs/Radio", () => "radio-component")

describe("RadioGroup", () => {
  const props: ComponentProps = {
    groupId: "fruit-radiogroup",
    label: "Select a fruit",
    labelId: "fruit-radiogroup-label",
    name: "fruit",
    onChange: jest.fn(),
    options: [
      {
        label: "Apple",
        value: "apple"
      },
      {
        label: "Banana",
        value: "banana"
      }
    ],
    selected: "banana"
  }

  it("renders", () => {
    const { container } = render(<RadioGroup {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })
})
