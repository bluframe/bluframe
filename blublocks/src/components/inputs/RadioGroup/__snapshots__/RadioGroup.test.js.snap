// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders 1`] = `
<div
  aria-labelledby="fruit-radiogroup-label"
  class="styled__Wrapper-sc-129dihy-0 hKQxru"
  id="fruit-radiogroup"
  role="radiogroup"
>
  <label
    class="styled__Label-sc-129dihy-1 ljQWBy"
    id="fruit-radiogroup-label"
  >
    Select a fruit
  </label>
  <div
    class="styled__Options-sc-129dihy-2 gSCsOv"
  >
    <radio-component
      inputid="fruit-radiogroup-apple"
      ischecked="false"
      label="Apple"
      name="fruit"
      value="apple"
    />
    <radio-component
      inputid="fruit-radiogroup-banana"
      ischecked="true"
      label="Banana"
      name="fruit"
      value="banana"
    />
  </div>
</div>
`;
